### PR TITLE
Parallel http requests

### PR DESCRIPTION
**Pull request by pjaspers**
## Why?

While publishing a rather large invoice, I found out that most of the time was spent in publishing to DHIS2.

When sending 11040 values to dhis2:

- Single request:    43 seconds (43093.83299993351)
- Parallel requests:  8 seconds (8223.681999836117)

Sending 3046 values had these results:

- Single request:  12 seconds
- Parallel rquests: 6 seconds

So big wins all around. I've also checked this on the DHIS2 side to confirm that the data is arriving.

``` sql      
select datavalue.lastupdated from datavalue 
  join organisationunit ON organisationunit.organisationunitid = datavalue.sourceid 
  join dataelement ON dataelement.dataelementid = datavalue.dataelementid 
where 
  dataelement.uid = 'WQbeZuIvhzS' and organisationunit.uid = 'XJkQtdGRO5Q';
```

And it was updating the lastupdated, it was also growing the audit table by the expected amounts.

## How?

`ParallelDhis2` uses the `typhoeus` gem to be able to make multiple concurrent requests to DHIS2, which is faster than the one huge request. ParallelDhis2 tries to mimick as much of its behavior from the normal `Dhis2::Client`, so that changes to that client or configuration are automatically taken over by ParallelDhis2.

### What's with the one huge file?

1. It's basically one class with two helper classes
2. It has a potential to be reused in other projects, so one file is
quicker to test out than a couple of files
3. I thought it was going to be smaller

## Production?

At the moment this feature is behind a feature flag (feature flags, are handled by `flipper` another gem I pulled in, to help with feature flags). The idea is to enable it for one specific `ProjectAnchor` so we can monitor the performance and results before enabling it by default.

### How to enable?

Either a console: `Flipper[:use_parallel_publishing].enable?(ProjectAnchor.find(<your-id_here>)`

Or visit `/flipper` in production and enable it using the UI (it's password protected).

## Aftermath

I've made a small adjustment to the output of the Dhis2Log that gets written by `ParallelDhis` runs, in the description will be `[parallel]` at the end. This is a small way to visually check who did the publishing.

The following snippet should paint a picture on whether it's working or not.

``` ruby
table = InvoicingJob.arel_table
anchor_id = <your_id_here>
before = InvoicingJob.where(project_anchor_id: anchor_id).where(table[:created_at].lt("20190214")).where.not(table[:duration_ms].eq(nil)).average(:duration_ms).to_f
after = InvoicingJob.where(project_anchor_id: anchor_id).where(table[:created_at].gt("20190214")).where.not(table[:duration_ms].eq(nil)).average(:duration_ms).to_f
puts "Before #{before} and After #{after}"
```

## Self proof reading checklist

- [x] Did I run Pronto and fixed all warnings (pronto run -c origin/dev)
- [x] Make sure all naming and code will remain understandable in 6 month by someone new to the project or by you.
- [x] Did I test the right thing?
- [x] Did I test corner cases, non happy path (defensive testing)?
- [x] Check that I used the ad-hoc patterns and created files accordingly (Presenters, Services, Search object, Form object,...)
- [x] Check code efficiency with record intensive project
- [x] Update documentation,readme accordingly

Thanks!

### CODE REVIEW

File: spec/lib/parallel_dhis2_spec.rb [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    debug:    true
+  }
+  options = default_options.merge(options)
+  Dhis2::Client.new(options)
+end
+
+def fake_response(status:, conflicts: [], description: "", import_count: {})
+  {
+    "status"            => status,
+    "conflicts"         => conflicts,
```
Comment (mestachs) [inline spec/lib/parallel_dhis2_spec.rb]: Can you make the fake_response more representative ?

when everything is fine, there's no conflicts field at all 

```
{
  "status": "SUCCESS",
  "description": "Import process completed successfully",
  "import_count": {
    "deleted": 0,
    "ignored": 0,
    "updated": 2133,
    "imported": 0
  },
  "response_type": "ImportSummary",
  "import_options": {
    "async": false,
    "dry_run": false,
    "sharing": false,
    "id_schemes": {
    },
    "merge_mode": "REPLACE",
    "report_mode": "FULL",
    "strict_periods": false,
    "import_strategy": "CREATE_AND_UPDATE",
    "skip_notifications": false,
    "skip_existing_check": false,
    "dataset_allows_periods": false,
    "skip_pattern_validation": false,
    "strict_organisation_units": false,
    "require_category_option_combo": false,
    "strict_category_option_combos": false,
    "require_attribute_option_combo": false,
    "strict_attribute_option_combos": false
  },
  "data_set_complete": "false"
}
```

can you double check when the status is error too ?

File: app/services/invoicing/invoice_entity.rb [hunk 2]
```diff
@@ -52,9 +56,9 @@ def fetch_and_solve
             invoicing_request.year_quarter.to_dhis2,
             solve_options
           )
-          @pyramid = fetch_and_solve.pyramid
-          @dhis2_export_values = fetch_and_solve.call
-          @dhis2_input_values = fetch_and_solve.dhis2_values
+          @pyramid = @fetch_and_solve.pyramid
```
Comment (pjaspers) [inline app/services/invoicing/invoice_entity.rb]: @mestachs I don't actually know how ruby handled this to be honest. But all tests are passing so I think this was the right move.

File: spec/lib/parallel_dhis2_spec.rb [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        client.post_data_value_sets(all_values)
+      }.to raise_error ParallelDhis2::TimedOut
+    end
+
+    it "raises on 500" do
+      stub_request(:any, expected_url).to_return(status: [500, "Internal Server Error"])
+      all_values = [{ value: 1234 }]
+      expect {
+        client.post_data_value_sets(all_values)
+      }.to raise_error ParallelDhis2::HttpException
```
Comment (mestachs) [inline spec/lib/parallel_dhis2_spec.rb]: from what I understand well, if we send 4 request and only 1 return get a 500, we won't log the partially sent values ?

Comment (pjaspers) [reply spec/lib/parallel_dhis2_spec.rb]: Correct. My train of thought was: if it's one big request that would fail we didn't log those values either, we'd just reran the request. So why make this exceptional if the standard is rerunning.

But I'm also not sure about that decision.

File: db/migrate/20190214095526_create_flipper_tables.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+class CreateFlipperTables < ActiveRecord::Migration[5.2]
+  def self.up
+    create_table :flipper_features do |t|
+      t.string :key, null: false
+      t.timestamps null: false
+    end
+    add_index :flipper_features, :key, unique: true
+
+    create_table :flipper_gates do |t|
+      t.string :feature_key, null: false
```
Comment (mestachs) [inline db/migrate/20190214095526_create_flipper_tables.rb]: don't know much about flipper  but the gates feature_key is the flipper_features.key ?
and in our case is use_parallel_publishing ?

Comment (pjaspers) [reply db/migrate/20190214095526_create_flipper_tables.rb]:
```
[scorpio_development] # select * from flipper_features;
 id |           key           |         created_at         |         updated_at         
----+-------------------------+----------------------------+----------------------------
  1 | use_parallel_publishing | 2019-02-14 10:15:58.267994 | 2019-02-14 10:15:58.267994
(1 row)

Time: 18.071 ms
[scorpio_development] # select * from flipper_gates;
 id |       feature_key       |  key   |      value       |         created_at         |         updated_at         
----+-------------------------+--------+------------------+----------------------------+----------------------------
  1 | use_parallel_publishing | actors | ProjectAnchor:15 | 2019-02-14 10:16:36.084829 | 2019-02-14 10:16:36.084829
```

File: spec/lib/parallel_dhis2_spec.rb (hunk 1 above)
Comment (pjaspers) [reply spec/lib/parallel_dhis2_spec.rb]: Good catch, both `SUCCESS` and `ERROR` didn't have it. I've now changed it so that I don't add it either.

File: app/services/invoicing/invoice_entity.rb (hunk 2 above)
Comment (pjaspers) [reply app/services/invoicing/invoice_entity.rb]: (The `@fetch_and_solve` is filled in, so when `fetch_and_solve` called it later in the method it would call itself, but the `@fetch_and_solve` would be memoized. I'm keeping this change because it feels more correct)

File: spec/lib/parallel_dhis2_spec.rb (hunk 3 above)
Comment (mestachs) [reply spec/lib/parallel_dhis2_spec.rb]: ok let's try like that

**Comment by pjaspers** [comment]
This turned out not be such a good idea:

```
table = InvoicingJob.arel_table
anchor_id = <your_id_here>
before = InvoicingJob.where(project_anchor_id: anchor_id).where(table[:created_at].lt("20190214")).where.not(table[:duration_ms].eq(nil)).average(:duration_ms).to_f
after = InvoicingJob.where(project_anchor_id: anchor_id).where(table[:created_at].gt("20190214")).where.not(table[:duration_ms].eq(nil)).average(:duration_ms).to_f
puts "Before #{before} and After #{after}"
```

The `InvoicingJob` gets reused, so the durations are overwritten, so this is not a good way to check if things are going better or worse. @mestachs would I be able to get it out of the DHIS2-logs?
